File: android/src/main/kotlin/dev/amphora/core/Reconciler.kt
package dev.amphora.core

import android.content.Context
import androidx.work.WorkInfo
import androidx.work.WorkManager
import dev.amphora.model.*
import dev.amphora.store.UploadDao
import dev.amphora.transport.TusTransport
import dev.amphora.work.UploadWorker

/**
 * The join between the two registries.
 *
 * Our Room database says what *should* be happening. WorkManager's own database — which survives
 * process death and reboot — says what *is*. Reconciliation is the intersection, and it is the
 * whole answer to "the user closed the app, reopened it, and expects to find their upload".
 *
 * Runs once per process start, before the host app is told the module is ready.
 * See docs/reference/persistence-and-recovery.md §5.
 */
class Reconciler(
    private val context: Context,
    private val dao: UploadDao,
    private val transport: TusTransport,
    private val engine: UploadEngine,
    private val storage: dev.amphora.governor.StorageGovernor,
    private val now: () -> Long = System::currentTimeMillis,
) {

    suspend fun reconcile(): ReconcileReport {
        var adopted = 0; var recovered = 0; var expired = 0; var failed = 0

        // 1–2. Working set, and free leases held by processes that no longer exist.
        dao.breakStaleLeases(now())
        val unfinished = dao.unfinished()

        // 3. What is genuinely still running?
        val live = liveWorkJobIds()

        for (job in unfinished) {
            if (job.id in live) {
                // ADOPT. Do not restart: a running worker plus a fresh one is two writers on one
                // upload URL, which corrupts silently. The lease would catch it, but not racing
                // in the first place is cheaper.
                adopted++
                continue
            }
            when (job.state) {
                UploadState.PAUSED -> Unit                       // sticky by design
                UploadState.BLOCKED -> engine.reevaluateGates(job)
                else -> { recoverOne(job).let { r -> when (r) {
                    Recovery.EXPIRED -> expired++
                    Recovery.FAILED -> failed++
                    Recovery.RESUMABLE -> recovered++
                    Recovery.DEFERRED -> Unit
                } } }
            }
        }

        // 5. Crash-path cleanup (state-machine I5).
        val orphanedFiles = storage.collectOrphans(dao.liveStagedPaths().toSet())
        retryPendingTerminations()

        return ReconcileReport(adopted, recovered, expired, failed, orphanedFiles)
    }

    private suspend fun recoverOne(job: UploadJob): Recovery {
        engine.dispatch(job.id, UploadEvent.ProcessStart)   // → RECOVERING

        // a. Is the source still there and unchanged?
        if (!engine.sourceIsIntact(job)) {
            engine.dispatch(job.id, UploadEvent.SourceMissing); return Recovery.FAILED
        }
        // b. Never created remotely — nothing to resume, start clean.
        if (job.uploadUrl == null) {
            engine.dispatch(job.id, UploadEvent.Retry); return Recovery.RESUMABLE
        }
        // c. Server-side lifetime elapsed.
        if (job.uploadExpiresAt != null && job.uploadExpiresAt < now()) {
            engine.dispatch(job.id, UploadEvent.Gone); return Recovery.EXPIRED
        }
        // d. Ask the only authority there is.
        return try {
            val head = transport.head(job.uploadUrl)
            when {
                head.offset < job.serverOffset -> {            // I7: never resume downward
                    engine.dispatch(job.id, UploadEvent.OffsetDiverged(head.offset)); Recovery.EXPIRED
                }
                else -> {
                    engine.dispatch(job.id, UploadEvent.OffsetAdvanced(head.offset)); Recovery.RESUMABLE
                }
            }
        } catch (e: Exception) {
            // An offline launch must still list and cancel jobs; only *resuming* needs network.
            engine.dispatch(job.id, UploadEvent.Blocked(BlockReason.NETWORK_UNAVAILABLE))
            Recovery.DEFERRED
        }
    }

    /** Cancelations whose DELETE never landed, because the device was offline at the time. */
    private suspend fun retryPendingTerminations() {
        for (job in dao.pendingTerminations()) {
            val url = job.uploadUrl ?: continue
            runCatching { transport.terminate(url) }
                .onSuccess { dao.update(job.copy(remoteTerminated = true, updatedAt = now())) }
        }
    }

    private fun liveWorkJobIds(): Set<String> =
        WorkManager.getInstance(context)
            .getWorkInfosByTag(WORK_TAG_PREFIX).get()
            .filter { it.state == WorkInfo.State.RUNNING || it.state == WorkInfo.State.ENQUEUED }
            .flatMap { info -> info.tags.mapNotNull { it.removePrefixOrNull("$WORK_TAG_PREFIX:") } }
            .toSet()

    private fun String.removePrefixOrNull(p: String) = if (startsWith(p)) removePrefix(p) else null

    private enum class Recovery { RESUMABLE, EXPIRED, FAILED, DEFERRED }

    companion object { private const val WORK_TAG_PREFIX = "amphora-job" }
}

data class ReconcileReport(
    val adopted: Int,
    val recovered: Int,
    val expired: Int,
    val failed: Int,
    val orphanedFilesRemoved: Int,
)

File: ios/Sources/Amphora/Core/Reconciler.swift
import Foundation
import os

/// The join between the two registries.
///
/// Our store says what *should* be happening. The background `URLSession` — whose tasks survive
/// app termination entirely — says what *is*. Reconciliation is the intersection, and it is the
/// whole answer to "the user closed the app, reopened it, and expects to find their upload".
///
/// Runs once per process start, before the host app is told the module is ready.
/// See docs/reference/persistence-and-recovery.md §5.
public actor Reconciler {

    private let log = Logger(subsystem: "dev.amphora", category: "reconciler")
    private let store: any UploadStore
    private let sessionManager: BackgroundSessionManager
    private let transport: any UploadTransport
    private let engine: any UploadEngine
    private let storage: StorageGovernor
    private let now: @Sendable () -> Date

    public init(
        store: any UploadStore,
        sessionManager: BackgroundSessionManager,
        transport: any UploadTransport,
        engine: any UploadEngine,
        storage: StorageGovernor,
        now: @escaping @Sendable () -> Date = Date.init
    ) {
        self.store = store
        self.sessionManager = sessionManager
        self.transport = transport
        self.engine = engine
        self.storage = storage
        self.now = now
    }

    public func reconcile() async -> ReconcileReport {
        var adopted = 0, recovered = 0, expired = 0, failed = 0

        // 1–2. Working set, and leases held by a process that no longer exists.
        try? await store.breakStaleLeases(asOf: now())
        let unfinished = (try? await store.unfinished()) ?? []

        // 3. What is genuinely still running? On iOS this is the strong case: tasks survive
        //    termination, so a job we left mid-transfer is very often still going.
        let live = await sessionManager.liveTasks()
        let liveByJob = Dictionary(uniqueKeysWithValues: live.map { ($0.jobId, $0) })

        for job in unfinished {
            if let task = liveByJob[job.id], task.state == .running || task.state == .suspended {
                // ADOPT. Do not restart: a live task plus a fresh one is two writers on one upload
                // URL, which corrupts silently. The lease (I8) would catch it, but not racing in
                // the first place is cheaper and avoids a wasted round trip.
                adopted += 1
                try? await store.update(id: job.id) { $0.taskIdentifier = task.taskIdentifier }
                log.debug("adopted live task for job \(job.id)")
                continue
            }

            switch job.state {
            case .paused:
                continue                                   // sticky by design
            case .blocked:
                await engine.reevaluateGates(job)
            default:
                switch await recover(job) {
                case .expired: expired += 1
                case .failed: failed += 1
                case .resumable: recovered += 1
                case .deferred: break
                }
            }
        }

        // 5. Crash-path cleanup (state machine I5).
        let livePaths = Set((try? await store.liveStagedPaths()) ?? [])
        let orphans = (try? await storage.collectOrphans(liveStagedPaths: livePaths)) ?? 0
        await retryPendingTerminations()

        return ReconcileReport(
            adopted: adopted, recovered: recovered, expired: expired,
            failed: failed, orphanedFilesRemoved: orphans
        )
    }

    private func recover(_ job: UploadJob) async -> Recovery {
        await engine.dispatch(jobId: job.id, event: .processStart)      // → .recovering

        // a. Is the source still there and unchanged?
        guard await engine.sourceIsIntact(job) else {
            await engine.dispatch(jobId: job.id, event: .sourceMissing)
            return .failed
        }
        // b. Never created remotely — nothing to resume, start clean.
        guard let uploadUrl = job.uploadUrl else {
            await engine.dispatch(jobId: job.id, event: .retry)
            return .resumable
        }
        // c. Server-side lifetime elapsed.
        if let expiresAt = job.uploadExpiresAt, expiresAt < now() {
            await engine.dispatch(jobId: job.id, event: .gone)
            return .expired
        }
        // d. Ask the only authority there is.
        do {
            let head = try await transport.head(uploadUrl: uploadUrl)
            if head.offset < job.serverOffset {              // I7: never resume downward
                await engine.dispatch(jobId: job.id, event: .offsetDiverged(serverOffset: head.offset))
                return .expired
            }
            await engine.dispatch(jobId: job.id, event: .offsetAdvanced(serverOffset: head.offset))
            return .resumable
        } catch {
            // An offline launch must still list and cancel jobs; only *resuming* needs network.
            await engine.dispatch(jobId: job.id, event: .blocked(.networkUnavailable))
            return .deferred
        }
    }

    /// Cancelations whose DELETE never landed, because the device was offline at the time.
    private func retryPendingTerminations() async {
        let pending = (try? await store.pendingTerminations()) ?? []
        for job in pending {
            guard let url = job.uploadUrl else { continue }
            if (try? await transport.terminate(uploadUrl: url)) != nil {
                try? await store.update(id: job.id) { $0.remoteTerminated = true }
            }
        }
    }

    private enum Recovery { case resumable, expired, failed, deferred }
}

public struct ReconcileReport: Sendable, Codable {
    public let adopted: Int
    public let recovered: Int
    public let expired: Int
    public let failed: Int
    public let orphanedFilesRemoved: Int
}
